modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java

```diff
@@ -0,0 +1,327 @@
+import org.apache.ignite.internal.catalog.events.DropIndexEventParameters;
+import org.apache.ignite.internal.catalog.events.MakeIndexAvailableEventParameters;
+import org.apache.ignite.internal.close.ManuallyCloseable;
+import org.apache.ignite.internal.lang.ByteArray;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.metastorage.Entry;
+import org.apache.ignite.internal.metastorage.MetaStorageManager;
+import org.apache.ignite.internal.metastorage.WatchEvent;
+import org.apache.ignite.internal.metastorage.WatchListener;
+import org.apache.ignite.internal.metastorage.dsl.Operation;
+import org.apache.ignite.internal.metastorage.dsl.Operations;
+import org.apache.ignite.internal.table.distributed.index.IndexBuildCompletionListener;
+import org.apache.ignite.internal.table.distributed.index.IndexBuilder;
+import org.apache.ignite.internal.util.Cursor;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+
+/**
+ * Component is responsible for ensuring that the index, upon completion of the distributed index building for all partitions, becomes
+ * available for read-write.
+ *
+ * <p>An approximate algorithm for making an index available for read-write:</p>
+ * <ul>
+ *     <li>On {@link CatalogEvent#INDEX_CREATE}, keys are created in the metastore: {@code startBuildIndex.<indexId>} and
+ *     {@code partitionBuildIndex.<indexId>.<partitionId_0>}...{@code partitionBuildIndex.<indexId>.<partitionId_N>}.</li>
+ *     <li>Then it is expected that the distributed index building event will be triggered for all partitions via
+ *     {@link IndexBuildCompletionListener} (from {@link IndexBuilder#listen}); as a result of each of these events, the corresponding key
+ *     {@code partitionBuildIndex.<indexId>.<partitionId>} will be deleted from metastore.</li>
+ *     <li>When all the {@code partitionBuildIndex.<indexId>.<partitionId>} keys in the metastore are deleted,
+ *     {@link MakeIndexAvailableCommand} will be executed for the corresponding index.</li>
+ *     <li>At {@link CatalogEvent#INDEX_AVAILABLE}, key {@code startBuildIndex.<indexId>} in the metastore will be deleted.</li>
+ * </ul>
+ *
+ * <p>Notes: At {@link CatalogEvent#INDEX_DROP}, the keys in the metastore are deleted: {@code startBuildIndex.<indexId>} and
+ * {@code partitionBuildIndex.<indexId>.<partitionId_0>}...{@code partitionBuildIndex.<indexId>.<partitionId_N>}.</p>
+ *
+ * @see CatalogIndexDescriptor#writeOnly()
+ */
+// TODO: IGNITE-20637 Recovery needs to be implemented
+// TODO: IGNITE-20637 Need integration with the IgniteImpl
+public class IndexAvailabilityController implements ManuallyCloseable {
+    private static final IgniteLogger LOG = Loggers.forClass(IndexAvailabilityController.class);
+
+    private static final String START_BUILD_INDEX_KEY_PREFIX = "startBuildIndex";
+
+    private static final String PARTITION_BUILD_INDEX_KEY_PREFIX = "partitionBuildIndex";
```

Comment:
It seems that if we include the period '.' in the prefix, the code will become simpler both for building and parsion the keys

Reply:
I do not think so =)

Reply:
While creating the full key, you will need to write `PREFIX + indexId` instead of `PREFIX + '.' + indexId`. This is shorter and simpler.

While parsing the full key, you will be able to omit the `indexOf('.', ...)` operation, so the code, again, becomes simpler.

And while registering watches on the prefix, the prefix with dot on the end will work the same.

2 of 3 operations will be made a bit simpler, for one of them nothing will change. There is a clear profit.

What are the benefits of the original code? Please elaborate.

Reply:
You're right, I'll fix it.